packages/styled-components/src/models/StyledComponent.js

```diff
@@ -40,212 +47,240 @@ function generateId(_ComponentStyle: Function, _displayName: string, parentCompo
+            `For example, { ${key}: () => InnerComponent } instead of { ${key}: InnerComponent }`
+        )
+    ),
+  };
+}
 
-      if (isFunction(resolvedAttrDef)) {
-        // $FlowFixMe
-        resolvedAttrDef = resolvedAttrDef(context);
-        attrDefWasFn = true;
-      }
+function useDevelopmentDeprecationWarnings(
+  displayName: ?string
+): $Call<typeof developmentDeprecationWarningsFactory, string> {
+  return useState(() => developmentDeprecationWarningsFactory(displayName))[0];
+}
 
-      /* eslint-disable guard-for-in */
+const useDeprecationWarnings =
+  process.env.NODE_ENV !== 'production'
+    ? useDevelopmentDeprecationWarnings
+    : // NOTE: return value must only be accessed in non-production, of course
       // $FlowFixMe
-      for (key in resolvedAttrDef) {
-        attr = resolvedAttrDef[key];
+      (() => {}: typeof useDevelopmentDeprecationWarnings);
 
-        if (!attrDefWasFn) {
-          if (isFunction(attr) && !isDerivedReactComponent(attr) && !isStyledComponent(attr)) {
-            if (process.env.NODE_ENV !== 'production') {
-              this.warnAttrsFnObjectKeyDeprecated(key, props.forwardedComponent.displayName);
-            }
+function useStyledComponentImpl<Config: {}, Instance>(
+  forwardedComponent: StyledComponentWrapper<Config, Instance>,
+  props: Config,
+  forwardedRef: Ref<any>
+) {
+  const {
+    attrs: componentAttrs,
+    componentStyle,
+    // $FlowFixMe
+    defaultProps,
+    displayName,
+    foldedComponentIds,
+    styledComponentId,
+    target,
+  } = forwardedComponent;
+
+  // NOTE: the non-hooks version only subscribes to this when !componentStyle.isStatic,
+  // but that'd be against the rules-of-hooks. We could be naughty and do it anyway as it
+  // should be an immutable value, but behave for now.
+  const theme = determineTheme((props: any), useContext(ThemeContext), defaultProps);
+
+  const utils = useDeprecationWarnings(displayName);
+  const [context, attrs] = useResolvedAttrs(theme, props, componentAttrs, utils);
+
+  const generatedClassName = useInjectedStyle(
+    componentStyle,
+    componentAttrs.length > 0,
+    context,
+    utils,
+    process.env.NODE_ENV !== 'production' ? forwardedComponent.warnTooManyClasses : (undefined: any)
+  );
 
-            attr = attr(context);
+  // NOTE: this has to be called unconditionally due to the rules of hooks
+  // it will just do nothing if it's not an in-browser development build
+  // NOTE2: there is no (supported) way to know if the wrapped component actually can
+  // receive refs -- just passing refs will trigger warnings on any function component child :(
+  // -- this also means we can't keep doing this check unless StyledComponent is itself a class.
+  //
+  // const refToForward = useCheckClassNameUsage(
```

Comment:
This is essentially not implementable with hooks, unless we make a dummy intermediate component **just** for calling `findDOMNode` in it, which sounds... icky. This warning was icky to begin with, as it triggers `StrictMode` warnings.

Reply:
plus, it has false positives, like for react-tooltip

Reply:
Yeah just delete it